### PR TITLE
Better icon image path detection

### LINKED ISSUE #6496: data-urls in L.Icon.Default
- [x] I've looked at the [documentation](http://leafletjs.com/reference.html) to make sure the behavior is documented and expected
- [x] I'm sure this is a Leaflet code issue, not an issue with my own code nor with the framework I'm using (Cordova, Ionic, Angular, React…)
- [x] I've searched through the issues to make sure it's not yet reported

**Steps to reproduce**

```
  var myIcon = L.Icon.Default.extend({options: {
    iconUrl: 'data:image/png;base64,iVBORw...'
  }});
```
**Expected behavior**
There is nothing special in [data url](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs), so I expect this code to work flawlessly.

**Current behavior**
Unfortunately there is path-guessing magic in code (#4605) which fails to deal with data-url.

**Environment**
- Leaflet version: 1.4.0
- Browser (with version): Google Chrome Version 71.0.3578.98 (Official Build) (32-bit)
- OS/Platform (with version): Windows 8.1 (32-bit)

**Additional context**
Current workaround:
```
  L.Icon.Default.imagePath = ' ';
```
But better to improve magic to take into account data-urls here:
https://github.com/Leaflet/Leaflet/blob/61ef1776ac3c7afc5dec5ce4dfab8986ff9f5376/src/layer/marker/Icon.Default.js#L45-L59

Or here:
https://github.com/Leaflet/Leaflet/blob/61ef1776ac3c7afc5dec5ce4dfab8986ff9f5376/src/layer/marker/Icon.Default.js#L33-L43

**Minimal example reproducing the issue**

I haven't got simple example, but will make if it will be necessary.

### PR DESCRIPTION
**Pull request by johnd0e**
Icon.Default/imagePath: more strict detect rules

Reject paths not ending on 'marker-icon.png'.
This should solve issues with images bundled into data-uri (webpack etc).

Add tests.

Close #6804 (superseded), fix #6496, fix (presumably) #7424.

### CODE REVIEW

File: src/layer/marker/Icon.Default.js [hunk 1]
```diff
@@ -57,6 +57,10 @@ export var IconDefault = Icon.extend({
 		           DomUtil.getStyle(el, 'backgroundImage');	// IE8
 
 		document.body.removeChild(el);
-		return this._stripUrl(path) || '';
+		path = this._stripUrl(path);
+		if (path) { return path; }
+		var link = document.querySelector('link[href$="leaflet.css"]');
```
Comment (IvanSanchez) [inline src/layer/marker/Icon.Default.js:62]: I guess this should be hoisted up, so if this runs successfully the fallback method doesn't run?

Comment (johnd0e) [reply src/layer/marker/Icon.Default.js:62]: I'm not sure: css-method can be used not only to detect path, it also allows to *override* path.

**Comment by simon04** [comment]
Various build systems (often by default)
- include asset files (only) when they are referenced in JS/CSS
- rename the asset files and include a hash in the filename. This facilitates caching (via `Cache-Control:immutable`)

For instance, bundling with [esbuild](https://github.com/evanw/esbuild) yields:
```bash
$ npm install leaflet esbuild
$ echo 'require("leaflet/dist/leaflet.css")' | node_modules/.bin/esbuild --bundle --outdir=dist/ --loader:.png=file
$ ls -1 dist/
layers-2x.CUVBMIZT.png
layers.ZHTVFDSJ.png
marker-icon.MCUQXS5S.png
stdin.css
stdin.js
```

`images/marker-icon.png` is renamed to `marker-icon.MCUQXS5S.png`, `images/marker-shadow.png` is missing.

The bundler webpack also includes a files hash by default, see [file-loader#options](https://webpack.js.org/loaders/file-loader/#options).

Does this PR take care of this de-facto standard behaviour of bundlers?

**Comment by johnd0e** [comment]
> Does this PR take care of this de-facto standard behaviour of bundlers?

It's impossible to guess random filenames, but you should be able to disable renaming in your bundler's config.
It would be wonderful if someone write small tutorial for such use case.

**Comment by simon04** [comment]
There's no need to guess the filenames. They are correctly referenced in the source files –

```css
$ grep -C2 png dist/stdin.css

.leaflet-control-layers-toggle {
  background-image: url(layers.ZHTVFDSJ.png);
  width: 36px;
  height: 36px;
}
.leaflet-retina .leaflet-control-layers-toggle {
  background-image: url(layers-2x.CUVBMIZT.png);
  background-size: 26px 26px;
}
.leaflet-default-icon-path {
  background-image: url(marker-icon.MCUQXS5S.png);
}
```

**Comment by johnd0e** [comment]
> They are correctly referenced in the source files –

That may be enough for layers, but javascript parts of ode still relies on original filenames - e.g. L.Marker, L.ImageOverlay.

**Comment by simon04** [comment]
Thus, the path-guessing heuristic should deal with `marker-icon-2x.png` and `marker-shadow.png` separately:
- `div.leaflet-default-icon-path` → `background-image` url → `L.Icon.Default.iconUrl`
- `div.leaflet-default-icon-path.leaflet-retina` → `background-image` url → `L.Icon.Default.iconRetinaUrl`
- `div.leaflet-default-icon-path.leaflet-marker-shadow` → `background-image` url → `L.Icon.Default.shadowUrl`

Related: https://github.com/Leaflet/Leaflet/pull/6951

**Comment by simon04** [comment]
@johnd0e, I'm happy to enhance my PR #6951 as suggested in https://github.com/Leaflet/Leaflet/pull/7092#issuecomment-724591326. What do you think, does this change have any chance to get merged?

**Comment by johnd0e** [comment]
> What do you think, does this change have any chance to get merged?

Sure, there are chances (but no promises).

**Comment by DimitrisRK** [comment]
Any news regarding this PR? I face trouble using default marker icons because of this issue and I prefer keeping webpack hash naming convention.

File: src/layer/marker/Icon.Default.js [hunk 2]
```diff
@@ -42,19 +42,25 @@ export var IconDefault = Icon.extend({
 		return (this.options.imagePath || IconDefault.imagePath) + Icon.prototype._getIconUrl.call(this, name);
 	},
 
+	_stripUrl: function (path) {	// separate function to use in tests
```
Comment (specious) [inline src/layer/marker/Icon.Default.js:45]: Might be more clear to say `// standalone function so it can be tested independently`

**Comment by specious** [comment]
It looks like a combination of #6951 and this PR might do the trick, however this is still a hack.

What exactly is the reason Leaflet needs to perform path manipulation to switch the marker image?

**Review by mourner — APPROVED** [review]

**Comment by mourner** [comment]
@Falke-Design nice work tracking all the relevant issues! Can you close those that no longer apply after merging this?

**Comment by simon04** [comment]
Still does not work 😞 

Here's a SSCCE based on the [esbuild](https://esbuild.github.io/) bundler: https://github.com/simon04/leaflet-esbuild/tree/leaflet-bug-7092 using the latest https://github.com/Leaflet/Leaflet/commit/80b268525e842fe087f0d058fd16efa75423557b. After running `esbuild --bundle --outdir=dist/ --loader:.png=file index.js` (via `yarn build`), the directory `dist/` contains the following PNG files, but Leaflet tries to load the non-existing `marker-icon-2x.png` –

```
dist/layers-55W3Q4RM.png
dist/layers-2x-TBM42ERR.png
dist/marker-icon-2V3QKKVC.png
```

**Comment by Falke-Design** [comment]
Like I understand this PR, it was not made to work with images names they contain hashs, it still needs `marker-icon.png`

**Pull request merged**